### PR TITLE
fix(http): classify request rejections instead of a blanket 400, and report uptime from /health

### PR DESCRIPTION
**Pull request by jmrplens**
## Description

The public endpoint answered `400 Bad Request` with the plain-text body `no server available` to any request without a `PRIVATE-TOKEN` header — including `initialize`. That string is not ours: go-sdk's streamable handler emits it whenever the `getServer` callback returns nil, which our HTTP handlers did for four unrelated reasons.

Two things were wrong beyond the lost detail:

1. **A missing credential is 401.** The MCP authorization spec's own error table says `401 — Authorization required or token invalid` and reserves `400` for a *malformed authorization request*. Nothing in the spec exempts `initialize` from authentication — the transport says servers **SHOULD** implement authentication for all connections — and protocol `2026-07-28` removed the initialize handshake entirely, so there is no handshake to open up.

2. **`400` now carries protocol meaning.** The `2026-07-28` transport tells a client that receives `400` with a body that is *not* a recognised JSON-RPC error to conclude the server is initialization-era and downgrade. A plain-text `no server available` therefore converted a missing header into a false protocol diagnosis — and the downgraded `initialize` `400`s too.

Observable cost: [chryaner/does-it-install](https://github.com/chryaner/does-it-install) records this endpoint as `connect_failed` ("endpoint unreachable") on darwin and linux. Its classifier (`src/harness/remote.ts:51`) maps `401`/`403` to `needs_auth` — "alive but requires credentials" — and every other HTTP status to `connect_failed`. Our `server.json` already declares `PRIVATE-TOKEN` as `isRequired: true`, so a `401` is enough to be recorded accurately.

This PR also adds process uptime to `/health` and bumps the patch version.

## Related Issue

N/A — found while reviewing the hosted endpoint's external reachability reports.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [x] Documentation update

### `fix(http)`: classify request rejections instead of a blanket 400

`mcpServerGate` resolves the pool server in a middleware ahead of the SDK handler and passes it through the request context, so each condition carries its own status:

| Condition | Before | After |
| --- | --- | --- |
| No `PRIVATE-TOKEN` / `Authorization: Bearer` | `400` `text/plain` `no server available` | `401` + `WWW-Authenticate` + JSON-RPC `-40100` |
| `GITLAB-URL` header not parseable | same opaque `400` | `400` + `-32600`, message names the header |
| Auth rate limiter exhausted | same opaque `400` | `429` + `Retry-After` + `-42900` |
| Pool/backend failure | same opaque `400` | `503` + `-50300`, internal error logged not returned |

- The legacy challenge is `Bearer realm="gitlab-mcp-server"` **without** `resource_metadata`. Clients discover an OAuth authorization server through that parameter and legacy mode has none (`/.well-known/oauth-protected-resource` is mounted only by `registerOAuthMCPHandlers`), so advertising it would start a discovery flow that cannot complete. `Bearer` is still honest — `ExtractToken` accepts `Authorization: Bearer <pat>` alongside `PRIVATE-TOKEN`. Per RFC 6750 §3.1 a challenge for a request carrying no credential omits the error code, so the accepted headers are named in the JSON message instead.
- All four bodies are JSON-RPC error responses with a null `id`, which the transport explicitly permits for rejections at this layer — and which is what stops a modern client from legacy-downgrading on the `400`.
- Error codes are allocated **outside** the JSON-RPC reserved range as the spec requires: `-32000..-32019` is legacy and `-32020..-32099` belongs to MCP. Ours mirror their HTTP status ×−100. A test enforces this.
- `GET`/`DELETE` are deliberately **not** gated — they keep reaching the SDK so it answers `405`, which is what `2026-07-28` prescribes.
- The same gate covers the OAuth path for the `GITLAB-URL` and pool cases, which previously fell through to the opaque `400` there too.

### `feat(http)`: report process uptime from `/health`

Adds `started_at` (RFC 3339, UTC) and `uptime_seconds` (whole seconds, truncated). Both, because they do different jobs: `started_at` is byte-identical across probes so a monitor can cache it and detect a restart by noticing it moved — the reason Prometheus exposes `process_start_time_seconds` rather than an uptime counter — while `uptime_seconds` is the derived convenience value, in the unit the [IETF health-check draft](https://datatracker.ietf.org/doc/html/draft-inadarei-api-health-check-06) uses for it (`"observedUnit": "s"`). RFC 3339 over epoch seconds matches how this project renders every other timestamp.

The arithmetic lives in `newHealthResponse(startedAt, now)` rather than reading the package clock: `TestHealthHandler_ReturnsOK` is already parallel, so a test mutating a package-level clock would race with it.

### `chore`: bump version to 2.6.6

`VERSION`, `CLAUDE.md` and the four distribution manifests; `llms*.txt`, `site/src/data/stats.json` and the README statistics section regenerated. `server.json` release-asset SHA256 hashes intentionally untouched — those are stamped after the tag by `scripts/update-server-json-sha.sh`.

### `docs(site)`: sync Starlight pages (EN + ES)

The site still mapped every failure to `400 / missing token header` and documented `tools/list` under "Health check". Both languages updated, including a row explaining that a plain-text `no server available` comes from the SDK in pre-2.6.6 builds and is not a crash.

## How to Test

1. Build and start in HTTP mode:
   ```bash
   go build -o /tmp/gls ./cmd/server
   /tmp/gls --http --http-addr=127.0.0.1:18080 --auto-update=false --tier=free --ignore-scopes
   ```
2. Unauthenticated `POST` is now a `401` with a challenge and a parseable body:
   ```bash
   curl -si -X POST http://127.0.0.1:18080/mcp -H 'Content-Type: application/json' \
     -H 'Accept: application/json, text/event-stream' \
     -d '{"jsonrpc":"2.0","id":1,"method":"initialize","params":{}}'
   # 401, WWW-Authenticate: Bearer realm="gitlab-mcp-server", application/json
   ```
3. A malformed instance URL names the header instead of going opaque:
   ```bash
   curl -s -X POST http://127.0.0.1:18080/mcp -H 'Content-Type: application/json' \
     -H 'Accept: application/json, text/event-stream' \
     -H 'PRIVATE-TOKEN: glpat-x' -H 'GITLAB-URL: ://nope' -d '{"jsonrpc":"2.0","id":1,"method":"initialize","params":{}}'
   # {"jsonrpc":"2.0","id":null,"error":{"code":-32600,"message":"Invalid GITLAB-URL header: malformed URL"}}
   ```
4. Eleven credential-less requests trip the limiter into `429` with `Retry-After: 60`.
5. `GET` still answers `405 Method Not Allowed` (`Allow: POST`).
6. `/health` reports uptime, with `started_at` stable across probes:
   ```bash
   curl -s http://127.0.0.1:18080/health
   # {"status":"ok",...,"started_at":"2026-08-22T19:43:06Z","uptime_seconds":9}
   ```

All six were run against the built binary, not only in tests.

## Breaking Changes / Migration Notes

No API or configuration change, but **HTTP status codes for failed requests change**. A client or monitor that treats `400` as "server needs a token" must now handle `401`; anything keying on the literal string `no server available` will no longer match. This is the intended correction — the previous behaviour misreported the cause.

### Code Quality

- [x] Code compiles: `go build ./...`
- [x] Consolidated Go analysis passes on changed packages
- [x] Code is formatted (`goimports`, `gofumpt`, `gci`)
- [x] Follows idiomatic Go patterns and the conventions in `CLAUDE.md` / `.github/instructions/`

### Testing

- [x] All existing tests pass: `go test ./cmd/server/ -count=1`
- [x] New tests added (15 total: 11 for the gate, 4 for uptime; table-driven)
- [x] Edge cases and error scenarios covered — all four rejection branches, non-POST pass-through, both accepted auth headers, pool built exactly once per token, truncation, clamping, UTC normalisation
- [x] Mutation-tested rather than assumed: disabling the gate reproduces the original `400`/`text/plain`/`no server available` and trips five assertions; switching uptime to milliseconds trips four; dropping `.UTC()` trips the format test
- [ ] E2E tests — not applicable, `test/e2e/suite/` uses in-memory transport and does not exercise the HTTP handler stack

### Documentation

- [x] Doc comments added for new types/functions
- [x] `docs/` updated (`http-server-mode.md`, `troubleshooting.md`)
- [x] Starlight site updated (EN + ES)
- [x] Commit messages follow Conventional Commits

### Security

- [x] No secrets, tokens, or credentials in code, tests, fixtures, or logs
- [x] Input validation for user-provided parameters
- [x] Error messages do not leak sensitive information — the pool error is logged but never returned; the `GITLAB-URL` parse detail is echoed only when the client actually sent the header, so a misconfigured server-side `--gitlab-url` is not reflected back
- [x] No new dependencies

## Follow-ups (not in this PR)

- **A bogus token still completes `initialize`.** The pool admits any non-empty string, so `PRIVATE-TOKEN: x` gets a full handshake and N distinct fake tokens churn the LRU. Format validation is not the answer — GitLab lets self-managed admins change the `glpat-` prefix — but `entryConfig` already performs authenticated I/O whose result we discard (`DetectScopes` returns `nil` on failure, and `ScopeSatisfied` reads `nil` as *allow all*). A `GET /user` admission check would be deterministic and format-agnostic.
- `--trusted-proxy-header` is confirmed set on the hosted deployment, so the `429` counts real client IPs there.

<a href="https://cubic.dev/pr/jmrplens/gitlab-mcp-server/pull/291?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by Sourcery

Classify HTTP authentication and backend failures accurately, add process uptime health metrics, and release version 2.6.6.

New Features:
- Expose process start time and whole-second uptime in the unauthenticated `/health` response.

Bug Fixes:
- Classify HTTP request failures with appropriate status codes and JSON-RPC error responses instead of returning an opaque blanket 400.
- Reject credentials explicitly refused by GitLab and prevent them from creating pooled sessions.

Enhancements:
- Preserve SDK method handling for non-POST requests while adding authentication and backend failure classification for MCP POST requests.

Build:
- Bump the project version and distribution manifests to 2.6.6.

Documentation:
- Update HTTP server, troubleshooting, and English and Spanish site documentation for rejection statuses, credential verification, and health metrics.

Tests:
- Add coverage for request rejection classification, credential admission, rate limiting, HTTP wiring, and health uptime calculations.

Chores:
- Regenerate versioned metadata and project statistics.

## Summary by CodeRabbit

- **New Features**
  - Added authentication and rate limiting for HTTP MCP requests.
  - Added clearer JSON-RPC responses for missing credentials, invalid URLs, rate limits, and server availability issues.
  - Added unauthenticated `GET /health` responses with status, version, startup time, and uptime.

- **Bug Fixes**
  - Improved request routing and reduced opaque server-availability errors.

- **Documentation**
  - Updated HTTP server and troubleshooting guidance, including health checks and authentication responses.

- **Chores**
  - Updated the release version to 2.6.6 and refreshed project statistics.

### CODE REVIEW

**Comment by jmrplens** [comment]
Pushed four more commits: the CI failure's root cause, the review feedback, and the credential admission check.

**CI failure — root cause found, and it is not this PR's code.** `TestServeHTTP_CrossOriginProtection_RejectsCrossSitePost` failed with `http server shutdown: context deadline exceeded`. It passed locally under four configurations: the package alone (×20), the package with the CI coverage flags, `GOMAXPROCS=1` with coverage, and the exact CI command `./cmd/... ./internal/...` pinned to 2 CPUs.

A controlled experiment settled it rather than inference: `http.Server.Shutdown` closes idle connections and then waits for active ones, and a connection whose response body the client has not consumed is not idle. Measured against a client holding an unread body, Shutdown takes exactly 5s and returns `context deadline exceeded` — the value of `httpShutdownTimeout` and the error CI reported. Drained first, it returns in 0s.

Every `serveHTTP` test used `defer resp.Body.Close()`, which runs at function exit, *after* `cancel()`. All of them raced Shutdown against a connection they still held open. Fast machines hide it because the client has already drained the small error body. Pre-existing latent flake in all of them; fixed at the source rather than by widening a timeout.

**Review feedback.** Applied: backend failures no longer consume auth-rate-limit capacity (a GitLab outage must not lock out valid tokens); `t.Run` subtests for the per-header cases; `io.ReadAll` instead of a single `Read`; corrected the `processStartTime` comment.

Checked and not applied, with reasons: `go run ./cmd/format_md_tables/ --check` exits 0 and the Analyze Markdown job passes, so the tables are not malformed; and the v2.6.6 release assets do not exist yet, so their SHA-256 values cannot be computed here — that stamping is what `scripts/update-server-json-sha.sh` does after the tag, which is why the 2.6.5 bump changed no `fileSha256` line either.

**Credential admission check** (the follow-up listed above, now implemented). The pool admitted any non-empty string, so `PRIVATE-TOKEN: x` completed a full handshake. `GET /user` is now asked once per new pooled session; only an explicit 401/403 rejects, mapping to 401 rather than 503. Everything else admits — failing closed on an unreachable GitLab would turn an outage into a denial of service.

Worth flagging: the first implementation used client-go and took the `serverpool` package from 0.5s to **150s**. retryablehttp with `RetryMax 5` turns a refused connection into seconds of backoff, which would have shipped a 5s stall on every new session against a struggling instance. Rewired through the non-retrying health client; the package is back to 0.77s.

File: cmd/server/authgate_test.go [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+		seen = nil
+		handler.ServeHTTP(rec, req)
+
+		if seen == nil {
+			t.Errorf("%s: handler received no server from the request context", header)
+		}
+		if rec.Code != http.StatusOK {
+			t.Errorf("%s: status = %d, want %d", header, rec.Code, http.StatusOK)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply cmd/server/authgate_test.go]: Applied in e01dfe40. Both cases are now `t.Run` subtests named after the header.

Kept `t.Parallel` out, as you flagged: the subtests share `seen`, and the `factoryCalls` assertion after the loop depends on both having run. A comment in the code now records that constraint so the next person does not add it back.

File: cmd/server/authgate_test.go [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+	resp, err := srv.Client().Do(req)
+	if err != nil {
+		t.Fatalf("request failed: %v", err)
+	}
+	defer resp.Body.Close()
+
+	body := make([]byte, 1024)
+	n, _ := resp.Body.Read(body)
+	got := string(body[:n])
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply cmd/server/authgate_test.go]: Applied in e01dfe40 — `io.ReadAll` with the error checked.

The reasoning holds beyond style: a short read would truncate the JSON and `decodeJSONRPCError` would fail with "body is not JSON", i.e. this end-to-end test would report a gate defect that does not exist. A test whose job is to rule out one specific regression must not have a second failure mode of its own.

File: cmd/server/authgate.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+		}
+		slog.Error("failed to create server for token", "error", err)
+		// The pool error can name internal state, so it is logged but not
+		// returned to the caller.
+		return nil, &gateFailure{
+			status:  http.StatusServiceUnavailable,
+			code:    errCodeUpstreamUnavailable,
+			message: "Could not initialize a GitLab session for this token. The instance may be unreachable; retry shortly.",
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply cmd/server/authgate.go:245]: Applied in e01dfe40. This was the substantive finding of the review.

A pool failure means the backend could not be reached or the server could not be built — the credential was never judged. Charging it to the per-IP limiter let a GitLab outage lock out clients holding valid tokens. That is the same conflation of unrelated causes this PR exists to remove, so leaving it was inconsistent with the change itself. (The previous code did it too; the classification just made it visible as a 429.)

One refinement worth recording for future reviews of this file: a190472c adds a case that genuinely *is* an authentication failure. When GitLab answers 401/403 to the credential probe, the pool returns `ErrInvalidCredential`, and the gate maps that to 401 and *does* record it against the limiter. So the rule is not "pool errors never count" — it is "only a verdict from GitLab about the credential counts". Errors where no verdict was obtained do not.

File: cmd/server/main.go [hunk 4]
```diff
@@ -1211,11 +1186,31 @@ func startPeriodicCleanup(ctx context.Context, cleanup func()) {
 	}()
 }
 
+// processStartTime marks when this process began serving.
+//
+// Package-level initialization runs before main, so this is the earliest
+// instant the program can observe about itself. Tests override it to make
+// uptime deterministic.
+var processStartTime = time.Now()
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply cmd/server/main.go:1195]: Applied in e01dfe40.

The comment was left over from an earlier draft that did override the clock. The final design deliberately does not: `newHealthResponse(startedAt, now)` takes both instants as parameters because `TestHealthHandler_ReturnsOK` is already parallel, so a test mutating a package-level clock would race with it. The comment now states that, which is the more useful fact — it explains why the seam exists.

File: docs/guides/http-server-mode.md [hunk 5]
```diff
@@ -253,12 +253,39 @@ If both headers are present, `PRIVATE-TOKEN` wins.
...
+#### Rejection Status Codes
+
+Every request that cannot be served is classified before it reaches the MCP handler, so the status distinguishes the cause:
+
+| Condition                                              | Status | JSON-RPC code | Notable headers    |
+| ------------------------------------------------------ | ------ | ------------- | ------------------ |
+| No `PRIVATE-TOKEN` and no `Authorization: Bearer`      | `401`  | `-40100`      | `WWW-Authenticate` |
+| `GITLAB-URL` header is not a parseable URL             | `400`  | `-32600`      | —                  |
+| More than 10 auth failures from one IP within a minute | `429`  | `-42900`      | `Retry-After`      |
+| GitLab session could not be built for the token        | `503`  | `-50300`      | —                  |
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply docs/guides/http-server-mode.md:284]: Not reproducible — leaving the tables as they are.

Verified twice, on the reviewed head and again after the latest push:

- `go run ./cmd/format_md_tables/ --check` exits 0 (`Markdown tables are up to date`). That is the exact gate the finding says would fail.
- The `Analyze Markdown` job passes on this PR.

The tables were run through `go run ./cmd/format_md_tables/` before committing, which is the documented step in CLAUDE.md for new or edited pipe tables.

For future reviews of this repo: cell padding here is normalised by that generator, not by markdownlint, and it pads each column to its widest cell — so header, separator and body rows are not expected to be identical in width across a table, and unequal raw widths are not on their own evidence of a formatting failure. The authoritative signal is the `--check` exit code. If there is a specific table where `--check` does fail, point me at it and I will fix it.

File: server.json [hunk 6]
```diff
@@ -31,8 +31,8 @@
   "packages": [
     {
       "registryType": "mcpb",
-      "identifier": "https://github.com/jmrplens/gitlab-mcp-server/releases/download/v2.6.5/gitlab-mcp-server-linux-amd64",
-      "version": "2.6.5",
+      "identifier": "https://github.com/jmrplens/gitlab-mcp-server/releases/download/v2.6.6/gitlab-mcp-server-linux-amd64",
+      "version": "2.6.6",
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply server.json:35]: Not actionable on this branch — the v2.6.6 assets do not exist yet.

`fileSha256` values are hashes of published release binaries. This branch only bumps the version; the release is cut later from the tag, and `scripts/update-server-json-sha.sh` then stamps the hashes back onto main from the release checksums file. There is nothing to hash at this point in the flow, so regenerating them here is not possible rather than merely skipped.

This is the repo’s normal release process, not an oversight in this PR — the previous bump (efc27778, v2.6.5) changed 0 `fileSha256` lines for the same reason, as does every bump before it. `make check-server-json` passes; it validates the schema, and hash freshness is a post-release property no pre-release commit can satisfy.

For future reviews: on this repository, a version bump that leaves `fileSha256` untouched is correct by design. Stale hashes are only a real finding on a commit made *after* the corresponding release exists.

File: internal/gitlab/client.go [hunk 7]
```diff
@@ -416,6 +416,35 @@
...
+// obtained and is reported as false, so callers fail open.
+func (c *Client) CredentialRejected(ctx context.Context) bool {
+	probeURL := strings.TrimRight(c.baseURL, "/") + "/api/v4/user"
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, probeURL, http.NoBody) //#nosec G704 -- built from a normalized GitLab base URL
+	if err != nil {
+		return false
+	}
+	req.Header.Set("PRIVATE-TOKEN", c.token)
+
+	resp, err := c.healthClient.Do(req) //#nosec G704 -- request URL derived from normalized GitLab config
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply internal/gitlab/client.go:438]: Dismissed as a false positive, matching the precedent already set for this exact rule in this exact file.

Alert #9 is `go/request-forgery` on `internal/gitlab/client.go:352` — the `Ping` request built from the same `baseURL` — and was dismissed for the same reason. The new alert is the same flow, not a new one.

Why it is accepted rather than fixed:

- The request target is the GitLab instance the operator pinned with `--gitlab-url`, or in multi-tenant HTTP mode the one the client selected via the `GITLAB-URL` header. Reaching a caller-selected instance is the entire purpose of that mode; every SDK call in the process already goes to the same host.
- The probe adds no network capability that did not already exist. `GET /api/v4/user` against that host was already issued by the scope and tier detection paths in `serverpool.entryConfig`, on the same code path, before this PR.
- The value is not raw. `serverpool.normalizeGitLabURL` rejects anything outside an `http`/`https` scheme, requires a host, and refuses embedded credentials, query strings and fragments before the URL ever reaches the client.

Also worth recording: this cannot be resolved by reshaping the code. Line 352 precomputes its URL into a struct field at construction time and is flagged anyway, so moving the probe URL out of the function body would not clear the rule — the taint is the configurable base URL itself, which is a product requirement.

The genuine mitigation would be an allowlist of permitted instance hosts for HTTP mode. That is a real feature with its own configuration surface, deliberately out of scope here, and tracked separately rather than smuggled into this PR.
